File: Codeforces/Codeforces Round #490 (Div. 3)/E - Reachability from the Capital.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 5000;

int n, m, s;
vector<int> g[N], gt[N], scc[N];
int u, v;
vector<bool> visited(N, false);
deque<int> stk;
vector<int> comp(N, -1);
int ctr = 0;
vector<int> indeg(N, 0);
vector<bool> seen(N, false);

void visit(int root) {
    if (!visited[root]) {
        visited[root] = true;
        for (auto nbr : g[root]) {
            visit(nbr);
        }
        stk.push_front(root);
    }
}

void assign(int u) {
    comp[u] = ctr;
    for (auto v : gt[u]) {
        if (comp[v] == -1) {
            assign(v);
        }
    }
}

int main() {
    cin >> n >> m >> s;
    s--;
    for (int i = 0; i < m; ++i) {
        cin >> u >> v;
        g[u - 1].push_back(v - 1);
        gt[v - 1].push_back(u - 1);
    }

    for (int i = 0; i < n; ++i) {
        visit(i);
    }

    for (auto u : stk) {
        if (comp[u] == -1) {
            assign(u);
            ctr++;
        }
    }

    for (int u = 0; u < n; ++u) {
        for (auto v : g[u]) {
            if (comp[u] != comp[v]) {
                scc[comp[u]].push_back(comp[v]);
                indeg[comp[v]]++;
            }
        }
    }

    stk.clear();
    stk.push_back(comp[s]);
    while (!stk.empty()) {
        int curr = stk.back();
        stk.pop_back();
        if (!seen[curr]) {
            seen[curr] = true;
            for (auto nbr : scc[curr]) {
                stk.push_back(nbr);
            }
        }
    }

    int ans = 0;
    for (int i = 0; i < ctr; ++i) {
        if (!seen[i] && indeg[i] == 0) {
            ans++;
        }
    }
    cout << ans << '\n';
    return 0;
}

File: Codeforces/Codeforces Round #490 (Div. 3)/E - Reachability from the Capital.py
from collections import deque

n, m, s = map(int, input().split())
g = {i: [] for i in range(1, n + 1)}
gt = {i: [] for i in range(1, n + 1)}
for i in range(m):
    u, v = map(int, input().split())
    g[u].append(v)
    gt[v].append(u)

visited = {i: False for i in range(1, n + 1)}
stk = deque()

def visit(root):
    if not visited[root]:
        visited[root] = True
        for nbr in g[root]:
            visit(nbr)
        stk.appendleft(root)

for i in range(1, n + 1):
    visit(i)

#print(stk)

component = {i: -1 for i in range(1, n + 1)}

def assign(u, root):
    if component[u] == -1:
        component[u] = root
        for v in gt[u]:
            assign(v, root)

for u in stk:
    assign(u, u)

#print(component)

component_indices = set(component.values())
scc = {i: [] for i in component_indices}
indeg = {i: 0 for i in component_indices}

for u in range(1, n + 1):
    for v in g[u]:
        if component[u] != component[v]:
            scc[component[u]].append(component[v])
            indeg[component[v]] += 1

#print(scc)
#print(indeg)

seen = {x: False for x in component_indices}
stk = [component[s]]
#print(stk)
while stk:
    curr = stk.pop()
    #print("curr: ", curr)
    if not seen[curr]:
        #print("Here")
        seen[curr] = True
        for adj in scc[curr]:
            stk.append(adj)

#print(seen)

ans = 0
for i in component_indices:
    if not seen[i] and indeg[i] == 0:
        ans += 1

print(ans)